resources/kernel/make_kernel_image_deb.sh

```diff
@@ -0,0 +1,127 @@
+# reference VMM.
+
+set -e
+
+SOURCE=$(readlink -f "$0")
+TEST_RESOURCE_DIR="$(dirname "$SOURCE")"
+
+trap "exit 1" TERM
+export TOP_PID=$$
+
+source "$TEST_RESOURCE_DIR/make_kernel.sh"
+
+KERNEL_VERSION="5.4.81"
+
+KERNEL_CFG="microvm-kernel-5.4-x86_64.config"
+
+# Reset index for cmdline arguments for the following `getopts`.
+OPTIND=1
+# Flag for optionally building a guest that halts.
+HALT=
+# Number of CPUs to use during the kernel build process.
+MAKEPROCS=1
+# Flag for optionally cleaning the workdir and recompiling the kernel.
+CLEAN=
+# Working directory. Defaults to a unique tmpdir.
+WORKDIR=$(mktemp -d)
+# Kernel binary format.
+KERNEL_FMT=
+# Destination kernel binary name.
+KERNEL_BINARY_NAME=
+
+USAGE="
+Usage: $(basename $SOURCE) -f (elf|bzimage) [-j nprocs] [-k kernel] [-w workdir] [-c] [-h]
+
+Options:
+  -f elf|bzimage    Kernel image format (either elf or bzimage).
+  -j nprocs         Number of CPUs to use for the kernel build.
+  -k kernel         Name of the resulting kernel image. Has the '-halt' suffix if '-h' is passed.
+  -w workdir        Working directory for the kernel build.
+  -c                Clean up the working directory after the build.
+  -h                Create a kernel image that halts immediately after boot.
+"
+export USAGE
+
+while getopts ":chf:j:k:w:" opt; do
+    case "$opt" in
+    c)  CLEAN=1
+        ;;
+    h)  HALT=1
+        ;;
+    f)  KERNEL_FMT=$(validate_kernel_format "$OPTARG")
+        ;;
+    j)  MAKEPROCS=$OPTARG
+        ;;
+    k)  KERNEL_BINARY_NAME=$OPTARG
+        ;;
+    w)  rm -rf "$WORKDIR"
+        WORKDIR=$OPTARG
+        ;;
+    *)  echo "$USAGE"
+        exit 1
+    esac
+done
+shift $((OPTIND-1))
+
+cleanup() {
+    if [ -n "$CLEAN" ]; then
+        echo "Cleaning $WORKDIR..."
+        rm -rf "$WORKDIR"
+    fi
+}
+
+# Step 0: clean up the workdir, if the user wants to.
+cleanup
+
+# Step 1: what are we building?
+[ -z "$KERNEL_BINARY_NAME" ] && KERNEL_BINARY_NAME=$(kernel_binary "$KERNEL_FMT")
+[ -n "$HALT" ] && KERNEL_BINARY_NAME="$KERNEL_BINARY_NAME-halt"
+
+# Step 2: start from scratch.
+mkdir -p "$WORKDIR" && cd "$WORKDIR"
+
+# Step 3: acquire kernel sources & config.
+extract_kernel_srcs "$KERNEL_VERSION"
+kernel_dir="$WORKDIR/linux-$KERNEL_VERSION"
+make_kernel_config "$TEST_RESOURCE_DIR/$KERNEL_CFG" "$kernel_dir"
+
+# Step 4: build *.deb packages.
+# We could build them from the kernel sources with `make deb-pkg`, but the
+# `deb-pkg` target forces a `make clean`, so we'll just download them instead
+# of building the entire kernel over and over again.
+# If, however, you want to build them from scratch, set the `MAKEDEB` env var.
+mkdir -p "$kernel_dir/deb"
+if [ -n "$MAKEDEB" ]; then
+    echo "Building deb packages..."
+    make_kernel "$kernel_dir" "$KERNEL_FMT" "deb-pkg" "$MAKEPROCS"
+    cp "$kernel_dir"/../*.deb "$kernel_dir/deb"
+else
+    echo "Downloading deb packages..."
+    DEB_URL="http://security.ubuntu.com/ubuntu/pool/main/l/linux-hwe-5.4"
```

Comment:
Hmmm, what are the deb packages for? It seems there are variables for things such as the kernel version, but then we have these links hard-coded here.

Reply:
It's the kernel image and modules. Even though we don't use the kernel shipped with the `linux-image` deb, Ubuntu still needs a bunch of modules in the `linux-modules` deb (and `System.map` for some reason) to work properly.

Reply:
As for the kernel version: we don't anywhere to download the *exact* kernel Ubuntu 20.04 ships with, so we instead build one with the same major.minor and thankfully it works fine. Turns out, you can't find *everything* on the Internet :(

Reply:
Hmm, aren't the modules and System.map available after building the kernel from sources as well? Also, does it make sense to use a kernel config where the required things are compiled as part of the image so we don't need to load modules?

Reply:
I think they can be built with other `make` recipes, but it takes way longer than it does to download them, and I'm not sure how to all the pieces come together to make a full Ubuntu. I went with this 2-step solution because it did the trick, and doesn't require multiple (re)compilations, but if you know a quicker/simpler way to get the working OS, we should go with it.

Reply:
I don't TBH, but the current one was/is a bit hard to understand and I was wondering if we can simplify things. I guess one way forward here is to go as things are until we have the initial devices merged and exercised by some tests as part of the CI etc, and then take a better look at the system itself, and at the very least add some more documentation if no other changes are required or important for clarity and reuse. What do other folks think here?

Reply:
We can go with how things are right now (the documentation is pretty clear) and, like you've said, think about simplifications after devices are tested by the CI. Also there is some duplicate code in the kernel scripts that could probably be avoided, but for now I'm okay with the PR.